### PR TITLE
Fixes bugs in stack unwinding metadata (PR#7118,7120)

### PR DESCRIPTION
**Pull request by btj**
This fixes a handful of bugs in ocamlopt's stack unwinding metadata generation. These have been causing crashes in my GUI application on OS X. (Cocoa performs stack walks as part of its normal operation. Incorrect stack unwinding metadata can cause it to try to read unmapped addresses. See PR#7118.) Besides, this also enables proper full stack backtraces in platform debuggers like gdb and lldb.

I added a regression test.

### CODE REVIEW

**Comment by gasche** [comment]
cc @xavierleroy @mshinwell 

(As noted in the Changes entry) this is the continuation of [PR#7118](http://caml.inria.fr/mantis/view.php?id=7118) (problem report) and [PR#7120](http://caml.inria.fr/mantis/view.php?id=7120) (problem analysis and fix suggestion).

**Comment by damiendoligez** [comment]
I'd like to see this reviewed by a back-end expert.

**Comment by mshinwell** [comment]
Acknowledged, I will look at this.
Has this been tested on x86-64 Linux?  If not please do so.

**Comment by btj** [comment]
So far, I only tested on OS X 10.10.1. (Travis tested on Linux, though.) I will test on Linux for any kind of regressions that I can think of checking for. But for this it would be useful to know what people are trying to achieve by emitting the CFI information.

**Comment by btj** [comment]
Answering my own question: CFI information was originally added on [Feb 21, 2012](https://github.com/ocaml/ocaml/commit/2eecf2d4c0933f64bcfa3e620f031497166db338) in response to [PR#5487](http://caml.inria.fr/mantis/view.php?id=5487), to get decent backtraces in gdb on Linux/x86 and Linux/amd64. I will test that this is still the case.

**Comment by mshinwell** [comment]
Right, the CFI information is so the stack can be unwound properly using a DWARF client such as gdb or lldb.  OCaml's internal backtrace infrastructure works from frame descriptors, and is independent.

**Comment by btj** [comment]
I built my fork (btj/ocaml) and ran the OCaml test suite on Ubuntu 15.10/amd64. All tests passed. I tested gdb backtrace behavior on a modified version of my regression test program (see [unwind.zip](https://github.com/ocaml/ocaml/files/108577/unwind.zip)); the [backtrace.txt](https://github.com/ocaml/ocaml/files/108581/backtrace.txt) looks fine.

**Comment by mshinwell** [comment]
@btj Did you test this with the OCaml compiler configured both with and without frame pointers (there is an option to the configure script)?  If not please do so (on both OS X and Linux if possible).

**Comment by btj** [comment]
Without frame pointers: see above.
With frame pointers:
- On ubuntu-14.04.3-desktop-amd64: builds fine, OCaml test suite passes, using my [unwind.zip](https://github.com/ocaml/ocaml/files/108577/unwind.zip) example I get a good-looking backtrace 
  [backtrace-with-fp.txt](https://github.com/ocaml/ocaml/files/125489/backtrace-with-fp.txt).
- On OS X 10.10.1: ./configure -with-frame-pointers reports an error, saying that frame pointers are not supported on this platform.

**Comment by damiendoligez** [comment]
Thank you for the patch and the tests.

**Comment by mshinwell** [comment]
I will finish looking at this shortly

**Comment by mshinwell** [comment]
@damiendoligez This patch is OK.  However could you please have a look and see whether the test Makefile is satisfactory?  For the moment this only does anything on OS X, but let's leave it like that for the moment, otherwise we need to add libunwind detection (which I have code for somewhere I can add for 4.04).  However I think it should be using $(OCAMLOPT) rather than just "ocamlopt".  I also wondered whether it should be changed so it doesn't clean itself every time.

@btj Thanks for the work on this.

File: testsuite/tests/unwind/Makefile [hunk 1]
```diff
@@ -0,0 +1,19 @@
+default:
+	@if $(BYTECODE_ONLY); then : ; elif [ $(SYSTEM) = macosx ]; then $(MAKE) native_macosx_tests; fi
```
Comment (damiendoligez) [inline testsuite/tests/unwind/Makefile]: This should be something like:

```
    @printf " ... testing 'unwind_test':"
    @if [ $(SYSTEM) = macosx ] && ! $(BYTECODE_ONLY); then \
      $(MAKE) native_macosx_test; \
    else \
      echo " => skipped"; \
    fi
```

... (see below)

File: testsuite/tests/unwind/Makefile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+native_macosx_tests:
+	@printf " ... testing 'unwind_test':"; \
```
Comment (damiendoligez) [inline testsuite/tests/unwind/Makefile]: ... and the `printf` should be removed from here.

File: testsuite/tests/unwind/Makefile [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        $(MAKE) clean ; $(MAKE) unwind_test && \
+	./unwind_test >/dev/null 2>&1 && echo " => passed" || echo " => failed"
+
+unwind_test:
+	@ocamlopt -c -opaque mylib.mli
```
Comment (damiendoligez) [inline testsuite/tests/unwind/Makefile]: @mshinwell is right, these should be `$(OCAMLOPT)` because you can't assume that OCaml is installed.

**Comment by damiendoligez** [comment]
@btj do you have the time to fix the tests quickly? If not, I'll do it myself and merge manually.

File: testsuite/tests/unwind/Makefile (hunk 3 above)
Comment (mshinwell) [inline testsuite/tests/unwind/Makefile]: (And, more importantly, if it is installed it is almost certainly not the compiler you want to test with!)

Comment (gasche) [inline testsuite/tests/unwind/Makefile]: (well, [opam-compiler-conf](https://github.com/gasche/opam-compiler-conf) makes it very easy to install a switch for the compiler branch you are currently hacking on.)

Comment (mshinwell) [inline testsuite/tests/unwind/Makefile]: The testsuite is always supposed to use the ocamlopt in the source tree, no?

**Comment by btj** [comment]
Done. $(OCAMLOPT): oops; yes, funny that I didn't notice that myself. I must have always run the test suite after installing the build into my PATH.

**Comment by damiendoligez** [comment]
On MacOS 10.8, the test fails with this error message:

```
ld: unknown option: -keep_dwarf_unwind
collect2: ld returned 1 exit status
```

The manual page for `ld` mentions `unwind` in only one place: the documentation of  `-warn-compact-unwind`.

I see two possible solutions:
1. Make the test conditional on the MacOS version (somehow).
2. Drop support for MacOS <= 10.8.

@btj which one would you recommend?

**Comment by btj** [comment]
I would recommend making the test conditional on the OS X version. I'll try to do that within the next few hours.

**Comment by btj** [comment]
Opened pull request #476.

**Pull request merged**